core/src/actors/epoch_manager/mod.rs

```diff
@@ -91,12 +114,169 @@ impl EpochManager {
-        info!(
-            "Checkpoint zero timestamp: {}",
-            self.checkpoint_zero_timestamp.unwrap()
+        debug!(
+            "Checkpoint zero timestamp: {}, checkpoints period: {}",
+            self.checkpoint_zero_timestamp.unwrap(),
+            self.checkpoints_period.unwrap()
+        );
+
+        // Start checkpoint monitoring process
+        self.checkpoint_monitor(ctx);
+    }
+    /// Method to compute time remaining to next checkpoint
+    fn time_to_next_checkpoint(&self) -> EpochResult<Duration> {
+        // FIXME(#145): Improve time precision, use nanoseconds
+        // Get current timestamp and epoch
+        let now = get_timestamp();
+        let current_epoch = self.epoch_at(now)?;
+
+        // Get timestamp for the start of next checkpoint
+        let next_checkpoint = self.epoch_timestamp(Epoch(
+            current_epoch
+                .0
+                .checked_add(1)
+                .ok_or(EpochManagerError::Overflow)?,
+        ))?;
+
+        // Get number of seconds remaining to the next checkpoint
+        let secs = next_checkpoint - now;
+
+        // Check if number of seconds to next checkpoint is valid
+        // This number should never be negative with current implementation
+        if secs < 0 {
+            Err(EpochManagerError::Overflow)
+        } else {
+            Ok(Duration::from_secs(secs as u64))
+        }
+    }
+    /// Method to monitor checkpoints and execute some actions on each
+    fn checkpoint_monitor(&self, ctx: &mut Context<Self>) {
+        // Wait until next checkpoint to execute the periodic function
+        ctx.run_later(
+            self.time_to_next_checkpoint().unwrap_or_else(|_| {
+                Duration::from_secs(u64::from(self.checkpoints_period.unwrap()))
+            }),
+            move |act, ctx| {
+                // Get current epoch
+                let current_epoch = match act.current_epoch() {
+                    Ok(epoch) => epoch,
+                    Err(_) => return,
+                };
+
+                // Send message to actors which subscribed to all epochs
+                for subscription in &mut act.subscriptions_all {
+                    subscription.send_notification(current_epoch);
+                }
+
+                // Get all the checkpoints that had some subscription but were skipped for some
+                // reason (process sent to background, checkpoint monitor process had no
+                // resources to execute in time...)
+                let epoch_checkpoints: Vec<_> = act
+                    .subscriptions_epoch
+                    .range(act.last_checked_epoch.unwrap_or(Epoch(0))..=current_epoch)
+                    .map(|(k, _v)| *k)
+                    .collect();
+
+                // Send notifications for skipped checkpoints for subscriptions to a particular
+                // epoch
+                // Notifications for skipped checkpoints are not sent for subscriptions to all
```

Comment:
Hmmm, we were not sure about this :sweat_smile: (we left a `TODO` for further discussion).

IMO, we should have the same "skipped epoch policy" for single-epoch notifications and for all-epoch notifications. That policy should be notifying for skipped epochs. And then, leave the logic of "what-to-do if past epoch notifications were received" to the actor that subscribed for those notifications.

Firstly, we were a bit worried about the actors subscribed to all epochs receiving too many epoch notifications if for any reason the `EpochManager` ticking mechanism was inhibited for a quite a long time. But I suppose that the subscribed actor could just discard them if this is a problem :thinking: 

Nevermind, if this functionality were to be implemented, the `EpochManager` would need to store the epoch at which each subscriber subscribed. This would be necessary to prevent `EpochManager` from notifying for epochs previous to the subscription time of each subscriber. We thought that if necessary, this extra implementation could be left to another issue to avoid further blocking.

What do you think about this? (Sorry for my "verbosity" :smile: )

Reply:
I agree with you, the changes needed to support such functionality may non-trivial and may add some unnecessary delay to this PR. I'm ok to leave it as is.